Preserve client certificate when TLS is indirectly enabled by a STS policy

progval opened the pull request:
Closes GH-4152

Only the last commit directly addresses the issue; the first four commits are only loosely related test refactoring (needed to test the fix) and additions (which test the basics this fix's tests rely on)

progval committed:
test/models/network.js: Fix test groupping

progval committed:
test/models/network.js: Reorder tests

progval committed:
test/models/network.js: Add a simple test for STS policies

progval committed:
test/models/network.js: Add tests for automatic client certificate creation/deletion

progval committed:
Preserve client certificate when TLS is indirectly enabled by a STS policy

Closes GH-4152.

MaxLeiter commented:
Looks good. My one potential change would be removing the [test/fixtures/env.js](https://github.com/thelounge/thelounge/pull/4477/files#diff-a0a3dc4f01870d9b985fe08b97efa26a8d7bc4e64d0bd20bb9c16a7dd8eff109) changes so the STS policy stuff is only enforced on the relevant tests .

progval commented:
I don't understand. Are you saying I should add

```
STSPolicies.refresh.cancel();
fs.unlinkSync(STSPolicies.stsFile);
```

to each of the three tests that use STS stuff?

MaxLeiter commented:
Yeah, maybe I'm misunderstanding but we only want that code to run after certain tests, right?

brunnre8 commented on test/models/network.js:
```diff
@@ -223,6 +223,28 @@ describe("Network", function () {
 			Helper.config.public = true;
 		});
+
+		it("should remove client certs if there is a STS policy", function () {
```
it **should remove** client certs?

>expect(ClientCertificate.get(network.uuid)).to.deep.equal(client_cert); // Should be unchanged

Well :wink: I think there's a not missing no?

brunnre8 commented:
Other than the test naming, this PR looks good I think.

But what I don't understand (and this has nothing to do with the PR itself, just poking your minds here).
Why on earth are we deleting the cert just because someone sets TLS to no TLS?

We shouldn't be **sending** the cert over plain text, but deleting it?
Say someone was doing troubleshooting over the plain text port, only to set it back to TLS and find their auth broken because the cert they setup is gone.

Anything I'm missing that warrants that behavior?

progval commented:
I completely agree. I assumed there was a good reason for this, but I'll gladly drop the cert removal entirely

progval committed:
Do not remove client certificate, even when TLS is disabled

It does not really make sense to remove it, as it can lock someone out of
their account, just by temporarily disabling TLS.

brunnre8 commented:
Thanks a lot Val :)
